### ISSUE TITLE
Lock file maintenance v2

### ISSUE THREAD

**Issue by rarkins** [renovateapp/renovate#1382]
### What is lock file "maintenance"?

It's when Renovate regenerates `yarn.lock` or `package-lock.json` files, in order to "refresh" them to latest for *all* dependencies, even though `package.json` hasn't changed.

### Why maintain lock files?

By their nature, lock files deliberately "hold back" updates to dependencies in the `package.json`. It's good to stop things breaking but bad if it means important fixes are held back.

### How does lock file maintenance work today?

Many don't use it (it's disable by default, due to many people being confused or irritated by it) and those that do usually schedule it for once a week.

### What's the idea for v2?

This idea is for repositories that have unpinned dependencies (i.e. use semver ranges in `package.json`). Using ranges is not always done because people want to cut down on PRs - it may be because they have no choice but to used ranged dependencies to avoid duplicates in downstream repositories.

If we parse the lock file, we can know exactly which version of a dependency is present there, e.g. the `package.json` range is `^1.5.0` but installed version in lock file is `1.6.4`. That way, by querying the npmjs registry we *know* immediately if the lock file needs to be updated, in the same way we know if pinned versions in a `package.json` need updating. So instead of upgrading everything once per week, we can upgrade dependencies individually and as needed. In a way, this treats the lock file as a secondary package definition and not as some type of "black box" binary.

This way there will be a new type of PR which contains changes to lock file only (like existing lock file maintenance PRs) but which is updating only one dependency and describes which one in the PR title and body, like we do with regular npm update PRs.

**Comment by rarkins** [comment]
@destroyerofbuilds and @ikatyang, I know you both use lock file maintenance currently - what do you think of this idea?

**Comment by destroyerofbuilds** [comment]
A large number of my personal project merge requests are [lockfile maintenance requests](https://gitlab.com/groups/hyper-expanse/-/merge_requests?state=merged).

> those that do usually schedule it for once a week.

Mine are scheduled for once per week.

> This idea is for repositories that have unpinned dependencies (i.e. use semver ranges in package.json). 

By default Renovate does not pin library dependencies, correct?

> Using ranges is not always done because people want to cut down on PRs 

I use ranges specifically to cut down on PRs (and also because pinning doesn't make sense. It doesn't respect the fact that transitive deps are not locked down in the same way).

> So instead of upgrading everything once per week, we can upgrade dependencies individually and as needed.

How are transitive dependencies updated in `yarn.lock`?

**Comment by ryansmith94** [comment]
Hey @rarkins, thought it might be useful for me to weigh in here.

So we have 4 "sub" (express) applications that can be running independent of each other (for scaling purposes), but we also have 1 (express) application that combines all four of those (into one express router) when we want to run them together.

In this case, the 4 sub applications need to use ranges so that the 1 combo app doesn't duplicate dependencies for the 4 sub applications. All 5 of these applications need their lockfile maintaining, but since we use the 4 sub applications on their own less often, I've only set lockfile maintenance on the combo application (scheduled once per week). This is mainly so that the lockfile is updated when there are patches to the 4 sub applications, but because it's once a week, patches to the 4 sub applications take an entire week to reach the combo application without manually updating the lockfile there. This means that I manually update the lockfile every time we patch the 4 sub applications.

All of this to say that in our case it would be really awesome if the lockfile maintenance was similar to the package file maintenance, so that patches to our 4 sub applications were immediately updated in the combo application. I hope that makes sense.

**Comment by felixfbecker** [comment]
https://www.dependencies.io/ supports this by having a multi-step process:
- Get all dependencies (including transitive) of the project
- For each, get the latest version
- Compare with current version
- If newer, run the package manager's update command to persist the latest version of the package to the lockfile

This makes some really nice workflows: https://github.com/felixfbecker/vscode-php-intellisense/blob/83846ffa0b22aebc30dd48131b7c40c54940b152/dependencies.yml#L13-L35

Because I can automatically pack and release a new patch release to the extension marketplace when their is e.g. a bug fix in the parser a (transitive dependency).

**Comment by felixfbecker** [comment]
@rarkins I think this is the number one most useful improvement for us - any ETA on when we would get to use this?

**Comment by rarkins** [comment]
@felixfbecker 	which package manager are you using now - npm or yarn?

**Comment by felixfbecker** [comment]
Awesome, will try it out!

The majority of projects use yarn now and prefer it for new projects.

**Comment by ljharb** [comment]
That’s a false statement in the general case, altho it may be true for you. npm is still the standard and the default, as well as the common case.

**Comment by felixfbecker** [comment]
Sorry I was replying to https://github.com/renovatebot/renovate/issues/1382#issuecomment-456173738

**Comment by felixfbecker** [comment]
@rarkins I enabled it yesterday on https://github.com/sourcegraph/lang-typescript but am not seeing any PRs even though there are a lot of in-range (lockfile) updates available

**Comment by rarkins** [comment]
It includes changes to the lockfile but appear to be fixing the http/https problem. I didn’t know that Yarn was affected by it too - thought it was only npm. I’ll look into the logs

**Comment by rarkins** [comment]
@felixfbecker there's 27 branches to create but 25 of them are in status "Skipping branch creation as not within schedule". If you add `"masterIssue": true` to your config then it would give you visibility of this type of thing, so I still recommend it.

**Comment by rarkins** [comment]
The PR was created as I think you might have a "no schedule" rule for internal packages, but then to make things even more confusing there's probably a bug in our semver resolution in this specific case *and* there happens to be `http` references in the `yarn.lock` too, which made Renovate think the file had changed and a PR was required. 

That package has some uncommon version suffixes, so I needed to use the online semver calculator myself to check that the current range is satisfied by only the one version:

![image](https://user-images.githubusercontent.com/6311784/51801076-5705f800-2239-11e9-8930-05fde947fd6a.png)

It does look like it needs upgrading:
![image](https://user-images.githubusercontent.com/6311784/51801105-a8ae8280-2239-11e9-8c4c-ee5ca32be022.png)

But that should be via a regular "replace" change to package.json and not a lockfile update alone, so I will troubleshoot it to work out the root cause. Right now I'm not sure why that would happen because we use npm's own semver package for resolutions.

**Comment by felixfbecker** [comment]
Yeah it makes sense that it would update because `w` comes after `f` and prereleases are sorted alphabetically. I am just confused why the PR doesn't actually contain the update.

**Comment by felixfbecker** [comment]
After disabling the schedule I am now getting PRs but they have this weird issue:

![image](https://user-images.githubusercontent.com/10532611/51801290-18be0800-223c-11e9-8ae7-364761bef8c0.png)

There are also unrelated updates contained in the PR: https://github.com/sourcegraph/sourcegraph-typescript/pull/84/files#diff-8ee2343978836a779dc9f8d6b794c3b2R837

This updates `@types/express` even though `ix` has no (transitive) dependency on it in any way.

**Comment by felixfbecker** [comment]
The update also seems to cause duplicate dependencies: https://github.com/sourcegraph/sourcegraph-typescript/pull/81/files#diff-8ee2343978836a779dc9f8d6b794c3b2R3374

This update of `fast-glob` at `^2.2.4` from `2.2.4` to `2.2.6` added `2.2.6` but still kept `2.2.4` around to satisfy `fast-glob@^2.0.2, fast-glob@^2.2.2`, which should be compatible with `2.2.6` and therefor deduplicated

**Comment by rarkins** [comment]
The duplication is a "characteristic" of `yarn` and would require post-processing of lock files using https://github.com/atlassian/yarn-deduplicate. Essentially, doing a simple `yarn add` or `yarn update` against an existing lock file does not mean that the whole tree gets shaken for duplicates.

**Comment by rarkins** [comment]
Re: the PRs with two commits, it seems they were done about 3 minutes apart, which is strange. Can you check the rebase box on one of them and see if the resulting force push from Renovate contains and retains only 1 commit?

**Comment by felixfbecker** [comment]
Interesting, I wasn't aware of that. That would be a good feature to have. I would imagine a flag to enable deduplication after an update.

**Comment by felixfbecker** [comment]
Does `lockfile-update` also update transitive dependencies? I don't see any PRs for that

**Comment by rarkins** [comment]
No, it doesn’t. It’s for updating direct dependencies only for now. But do you think your teams could accept the 100s of PRs per week it would likely generate? Otherwise lock file maintenance can update *all* in the lock files

**Comment by felixfbecker** [comment]
We have it configured to update only in the first week of each month. I found that the lock file maintenance PRs are just way to risky to merge - there are too many updates and too little insight into what changed. I don't see why direct dependency updates are more important than indirect dependency updates - imo all arguments for why direct dependencies should be updated and updated in separate branches apply equally to transitive dependencies. Some transitive dependencies we maintain ourselves so want to configure them with automerge / disable schedule etc. Maybe it would be too many PRs and we would need to configure more groups, prioritise them differently by importance of certain packages etc but drawing the line between direct dependencies and transitive dependencies is arbitrary, and currently transitive dependencies are excluded from `packageRules` so can't be configured at all.

**Comment by ljharb** [comment]
I think the main difference is that you chose your direct dependencies; their choice of transitive deps isn’t supposed to even have to enter your observation. The need to, or ability to, update transitive deps is generally because of encapsulation bugs in your directs.

**Comment by felixfbecker** [comment]
That argument would make sense if floating ranges didn't exist and/or weren't the standard way to define dependencies. Nobody releases a new (patch) version of their package with a bumped dependency because the dependency released a new (patch) version of their package.

**Comment by ljharb** [comment]
They absolutely do (and always should imo) when they become aware of a bug affecting their users.

**Comment by felixfbecker** [comment]
@ljharb if that was true, why do floating ranges even exist in the first place (and are the default for `npm install`)? There would be no point if it was truly feasible and the practice for every package to bump their own version and rerelease on every single update to their transitive dependencies to propagate fixes through the tree.

It's _good_ that end users have control. As an end user, I want the latest fixes asap, no matter if all the packages in the dependency chain bumped and rereleased (yet). And as a heavy user of Renovate I would love for Renovate to give me that option. You won't be forced to use it 😉 

As for the implementation, I think it should be relatively simple. `yarn upgrade` is able to upgrade transitive dependencies afaik, so it would just have to walk the dependency tree, check for new versions and run `yarn upgrade $PACKAGENAME` if available.

**Comment by ljharb** [comment]
@felixfbecker I think we're misunderstanding each other; I'm saying that floating ranges *do* exist, and when a transitive dependency updates with a nonbreaking change, it's *supposed* to filter to all dependents without any changes needed (except perhaps in a top-level lockfile).

I'm saying that the updates of your directs are what's important for you to review and consider; the updates to your transitives can theoretically just be rubber-stamped as long as your tests are passing.

**Comment by felixfbecker** [comment]
Okay - in theory, yes, but in practice you need to do a risk assessment on transitive dependency updates just as much as you need to do a risk assessment on direct updates. Note I am saying "just as much" - if you have perfect test coverage, then wouldn't need to assess direct updates either and could always automerge, but nobody has _perfect_ test coverage. We've been bitten by sneaky-ninja transitive updates in the past: https://about.sourcegraph.com/blog/the-pain-that-minimal-version-selection-solves#background
While in theory they are just an implementation detail, in reality they need to be given the same care as direct updates.
